queasars/job_shop_scheduling/domain_wall_hamiltonian_encoder.py

```diff
@@ -0,0 +1,528 @@
+        self._hamiltonian_prepared = True
+
+    def _operation_overlap_term(self, operation_1: Operation, operation_2: Operation, penalty: float) -> SparsePauliOp:
+        """
+        Return a SparsePauliOp which penalizes variable states which make operation_1 and operation_2 overlap.
+        Within a hamiltonian this term evaluates to zero if the two operations do not overlap and to penalty if they
+        overlap
+
+        :arg operation_1: operation which must not overlap operation_2
+        :type operation_1: Operation
+        :arg operation_2: operation which must not overlap operation_1
+        :type operation_2: Operation
+        :arg penalty: penalty value which is applied if the operations overlap
+        :type penalty: float
+        :return: a SparsePauliOp which penalizes variable states in which the two operations overlap
+        :rtype: SparsePauliOp
+        """
+        start_variable_1 = self._operation_start_variables[operation_1]
+        start_variable_2 = self._operation_start_variables[operation_2]
+
+        if start_variable_1.max_value + operation_1.processing_duration <= start_variable_2.min_value:
+            return 0 * _constant_one_term(n_qubits=self._n_qubits)
+
+        if start_variable_2.max_value + operation_2.processing_duration <= start_variable_1.min_value:
+            return 0 * _constant_one_term(n_qubits=self._n_qubits)
+
+        local_terms = []
+        for start_time_1 in start_variable_1.values:
+            for start_time_2 in start_variable_2.values:
+                if (
+                    start_time_1 <= start_time_2 < start_time_1 + operation_1.processing_duration
+                    or start_time_2 <= start_time_1 < start_time_2 + operation_2.processing_duration
+                ):
+                    local_terms.append(
+                        penalty
+                        * start_variable_1.value_term(
+                            value=start_time_1, quantum_circuit_n_qubits=self._n_qubits
+                        ).compose(
+                            start_variable_2.value_term(value=start_time_2, quantum_circuit_n_qubits=self._n_qubits)
+                        )
+                    )
+        return SparsePauliOp.sum(local_terms)
+
+    def _operation_precedence_term(
+        self, operation_1: Operation, operation_2: Operation, penalty: float
+    ) -> SparsePauliOp:
+        """
+        Returns a SparsePauliOp which penalizes variable states in which operation_2 start before operation_1 ends.
+        Within a hamiltonian this term evaluates to zero only if operation_2 starts after (or at the same time) the
+        operation_1 has ended. Otherwise, it evaluates to penalty
+
+        :arg operation_1: operation which must precede operation_2
+        :type operation_1: Operation
+        :arg operation_2: operation which must start after (or at the same time) operation_1 has finished
```

Comment:
"must" is a bit strong here, no? I mean it's not a requirement for this method but rather the effect the output of the method is supposed to have.

Reply:
Right. I changed this to use should